### PR TITLE
Change concat dims to be Hashable

### PR DESCRIPTION
**Pull request by max-sixty**
A small typing change

### CODE REVIEW

**Comment by mathause** [comment]
Can you do the other `overload` and the docstring? Did you test if it works passing `dim=(“x", "y")`?

**Comment by max-sixty** [comment]
I broadened this out into using `Literal` and adding annotations to almost all `test_concat` files

File: xarray/tests/test_concat.py [hunk 1]
```diff
@@ -50,10 +52,10 @@ def test_concat_compat():
 
 class TestConcatDataset:
     @pytest.fixture
-    def data(self):
+    def data(self) -> None:
```
Comment (Illviljan) [inline xarray/tests/test_concat.py:55]:
```suggestion
    def data(self) -> Dataset:
```
Seems to return a Dataset, right?

File: xarray/core/concat.py [hunk 2]
```diff
@@ -24,14 +25,19 @@
...
 def concat(
     objs: Iterable["Dataset"],
-    dim: Union[str, "DataArray", pd.Index],
-    data_vars: Union[str, List[str]] = "all",
-    coords: Union[str, List[str]] = "different",
-    compat: str = "equals",
+    dim: Union[Hashable, "DataArray", pd.Index],
+    data_vars: Union[concat_options, List[Hashable]] = "all",
+    coords: Union[concat_options, List[Hashable]] = "different",
+    compat: compat_options = "equals",
```
Comment (Illviljan) [inline xarray/core/concat.py:40]: How about removing the Union and using `|` while we're here?

```suggestion
    dim: Hashable | "DataArray" | pd.Index,
    data_vars: concat_options | List[Hashable] = "all",
    coords: concat_options | List[Hashable] = "different",
    compat: compat_options = "equals",
```

File: xarray/core/concat.py [hunk 3]
```diff
@@ -43,10 +49,10 @@ def concat(
...
 def concat(
     objs: Iterable["DataArray"],
-    dim: Union[str, "DataArray", pd.Index],
-    data_vars: Union[str, List[str]] = "all",
-    coords: Union[str, List[str]] = "different",
-    compat: str = "equals",
+    dim: Union[Hashable, "DataArray", pd.Index],
+    data_vars: Union[concat_options, List[Hashable]] = "all",
+    coords: Union[concat_options, List[Hashable]] = "different",
+    compat: compat_options = "equals",
```
Comment (Illviljan) [inline xarray/core/concat.py:55]: Same here.

```suggestion
    dim: Hashable | "DataArray" | pd.Index,
    data_vars: concat_options | List[Hashable] = "all",
    coords: concat_options | List[Hashable] = "different",
    compat: compat_options = "equals",
```

**Review by Illviljan — COMMENTED** [review]
Looking good. :)

File: xarray/core/concat.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/concat.py:40]: Yes, I thought `pyupgrade` did this for us; I can have a look why that isn't working if no one gets there first

**Comment by max-sixty** [comment]
Thanks @Illviljan !

File: xarray/core/concat.py [hunk 4]
```diff
@@ -12,6 +14,7 @@
 )
 
 import pandas as pd
+from typing_extensions import Literal
```
Comment (dopplershift) [inline xarray/core/concat.py:17]: I'm not sure why tests didn't fail here, but unconditionally importing from `typing_extensions` causes `import xarray` to fail on Python >=3.8.

Comment (Illviljan) [reply xarray/core/concat.py:17]: #6159 should fix this error. You might get errors for python 3.7 instead though, see #5892.

Comment (dopplershift) [reply xarray/core/concat.py:17]: We only run nightly with latest Python, so that's fine. I'll deal with the Py 3.7 removals another day...

**Pull request merged**